enhancements/machine-api/spot-instances.md

```diff
@@ -0,0 +1,459 @@
+
+- The actuator is responsible for creation of the instance only and should not attempt to remediate problems
+
+- The actuator should not attempt to verify that an instance can be created before attempting to create the instance
+  - If the cloud provider does not have capacity, the Machine Health Checker can (given required MHC) remove the Machine after a period.
+    MachineSet will ensure the correct number of Machines are created.
+
+- There’s a [2 hours time frame](https://github.com/openshift/cluster-machine-approver/blob/master/csr_check.go#L38) in OpenShift for a machine to become a node.
+  If the Spot request is not satisfied during this time frame, it is assumed that the instance will never become a node.
+  This machine would be remediated by an MHC if present.
+
+- Graceful termination of nodes should be provided by observing termination notices
+
+### Implementation Details
+
+#### Termination handler design
+
+To enable graceful termination of workloads running on non-guaranteed instances,
+three controllers will need to be implemented to watch for termination notices
+and gracefully move workloads
+
+##### Termination Pod DaemonSet
+
+The Termination Pod DaemonSet would be deployed will be deployed by the Machine API Operator and
+will select Nodes which are labelled as spot instances.
+This will ensure the Termination Pod only runs on instances that need it.
+
+The spot label can be added by the cloud provider actuators as they create instances,
+provided they support spot instances and the instance is a spot instance.
+
+##### Termination Pod
+
+The Termination Pod is the Pod that will be deployed to individual Nodes to
+monitor for termination notices for that Node.
+
+It will operate on the following logic:
+- Call actuator method to watch for termination notice
+  - This should return a channel which is to be closed when the termination notice is observed
+- Wait for the actuator signal channel to be closed
+- Taint the Node
+- Delete the Machine for the Node that the Pod is running on
+- Stop watching for further termination notices
+
+This will be a controller within the Machine API Operator repository and will
+be deployed by the Termination Pod DaemonSet.
+This will require a new method for each cloud provider actuator to watch for termination notices.
+
+###### Alternative draining
+
+As an alternative to directly deleting the Machine objects,
+the Termination Pod could instead add a condition to the Node object for Node that the Pod is running on.
+
+This condition could be used by an optional MachineHealthCheck which would monitor the Nodes for this condition.
+
+#### Cloud Provider Implementation Specifics
+
+##### AWS
+
+###### Launching instances
+
+To launch an instance as a Spot instance on AWS, a [SpotMarketOptions](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#SpotMarketOptions)
+needs to be added to the `RunInstancesInput`. Within this there are 3 options that matter:
+
```

Comment:
which particular fields would we be exposing at the user facing API?

Reply:
This is explained below on lines L#138-L#151